### ISSUE TITLE
A meta-proposal: add usability analysis to each language enhancement proposals

### ISSUE THREAD

**Issue by Emasoft** [dart-lang/dart_enhancement_proposals#13]
I have a meta-proposal about proposals: add usability analysis to each language enhancement proposals.

For usability analysis I mean something more than just making general usability considerations, like in some of the proposals I've read. Many of those proposals add great functionalities to the language, but they also introduce complexity and make the language more criptic and harder to grasp and use.

Each proposal should have a section labeled "Usability Analysis", where the following 10 question are each answered with a Yes or No, and each accompanied with an explanation:

1 - Is the code more or equally readable after the enhancement than before?
2 - Is the syntax of the enhancement similar to the same functionality in other languages, so that it appears already familiar to the majority of programmers?
3 - Is any new behaviour introduced by the enhancement accompanied by a new and unique recognizable keyword or character string, and not reusing an existing one that makes harder to distinguish it?
4 - Are the keywords, the identifiers or the character sequences used to implement the enhancement meaningful and unique?
5 - Are the syntax, identifiers and keywords introduced by the enhancement verbose and explicit enough to be expressive and clear about what it does and how it does it?
6 - Does an user that has never seen the enhancement syntax before, recognize or infer its functionality easily with no help but just reading the code?
7 -  Is there only one way to perform the introduced functionality, and no multiple syntaxes to obtain the same result?
8 - Is the enhancement designed to avoid potential pitfalls and involuntary bugs in its usage? Does the enhancement includes enough limitations and guards against the worst case scenarios to be considered fool proof?
9 - Can the code using the new functionality be read aloud as an explicit verbal statement and make perfect sense?
10 - Is the syntax of the new functionality context independent, and its functionality not different when the same syntax or identifier is used in a different context?

Each of those 10 questions adds 1 point to the "Usability Score" if answered positively.
An enhancement proposal should score at least 6 Usability Points to be considered viable.

**Comment by sethladd** [comment]
cc @lukechurch

**Comment by NDxTreme** [comment]
This sounds good at first glance. Maybe it can help refine my own thought process. Thanks.

**Comment by munificent** [comment]
I definitely agree that usability is a key metric for any proposal. I don't think it can be easily broken down to a simple numeric score. (For example, how would you define "more readable" on your first question.)

Maybe the best way to approach this is to add most of these as suggestions to consider in the proposal template?

**Comment by Emasoft** [comment]
@munificent You are right about the lack of a rigorous definition for readability and other elements that the 10 questions ask to evaluate. But the problem is that in the field of usability the science of reference is human psychology or at best neuroscience. The brain is a difficult thing to model mathematically, so it's very hard to define exactly what the brain find more or less readable. To solve the problem we need to ask to the brain itself each time, and at the same time trying to not be influenced by the preexisting memes in the brain. 
For example: is "foo++" more or less readable than "foo=foo+1"? Even if we know the meaning of the former, it is clear that the latter is the more intelligible, because is more explicit and does not require a custom symbol like "++", but only those of basic algebra. But at the same time using "foo++" in a long statement makes the code more compact and faster to grasp than a code fragmented by a lots of "foo=foo+1".
So you may say that:
READABILITY = INTELLIGIBILITY * COMPACTNESS
Both are factors, because if you have 0 intelligibility and 100 compactness, you still have an unreadable code. And if you have a syntax too much verbose and fragmented (0 compactness),even if you have 100 intelligibility (you write in plain english) you still get an unreadable code.
But, as I said, it is very hard to know in advance what the brain will find easier to read. We don't have a working model of the cognitive functions of the human brain yet, not even an approximate model. So we must in the end resort to asking to the brain itself. It may not be a very consistent and reliable assessment due to the influence of different preexisting memes in different subjects. But at least in making this usability assessment, the author of the proposal will be forced to take those 10 elements in consideration trying to reach a better usability score. And the reviewers of the proposal will be able to criticize the answers given to those 10 questions, reporting a different answer for some of the questions based on their subjective assessment. The result will be a much more structured and useful discussion on the usability aspects of the proposal.

**Comment by lukechurch** [comment]
@Emasoft @munificent @sethladd @gbracha 

I love the idea of adding a usability analysis to proposals. I think the most important part of this is that it would mandate that people supply an argument justifying their proposal in terms of usability, rather than seeking a number. 

The system being proposed here is known in UX circles as a form of 'heuristic evaluation'. Such heuristics themselves are hard to come up with and evaluate for efficacy and completeness. The scheme proposed here focuses primarily on readability, important certainly - but not sufficient. 

Even though we don't have a mathematical model for the brain, we don't need one to, as you say there are other disciplines like psychology that can help us here. There are people who have spent a couple of decades thinking about what makes notation systems usable, and some of their models have gone through some basic cross-validation. 

http://www.ppig.org/papers/12th-blackwell.pdf, (snippits extracted at the bottom) is probably the most immediately relevant set of questions. This is a scheme I've worked with extensively on other programming design projects and have generally found it very helpful.

My suggestion is that in order to be considered a proposal must include a 'usability analysis', which justifies the UX implications of the change. It is up to the authors of the proposal what they think is the most persuasive way of demonstrating UX benefits, however I am most likely to be persuaded of completeness of a UX proposal if it has answers to the below for the proposal. For some proposals, some of those answers may be "this needs tool support to do x", or "not applicable here".

If the proposal proceeds to prototype, in many cases it's reasonable to run empirical studies to determine whether it has e.g. readability problems or not. I'd rather we used validated heuristics for the proposal and focused on getting numbers measured from studies.

This may seem like a lot of work, but I think it's a fairly reasonable amount given the complexity of assessing the user impact of changes to a programming language. The questionnaire below often takes around an hour or so to fill out. Given that a change would [hopefully] positively improve every Dart developer's UX, that seems a fairly small cost.

Thanks for bringing this up, I think it's a very important aspect of the process.


### Viscosity
• When you need to make changes to previous work, how easy is it to make the change? Why?
• Are there particular changes that are more difficult or especially difficult to make? Which ones?

### Hidden Dependencies
• If the structure of the proposal means some parts are closely related to other parts, and changes to one may affect the other, are those dependencies visible? What kind of dependencies are hidden?
• In what ways can it get worse when you are creating a particularly large description?
• Do these dependencies stay the same, or are there some actions that cause them to get frozen? If so, what are they?

### Visibility and Juxtaposability
• How easy is it to see or find the various parts of the notation while it is being created or changed? Why?
• What kind of things are more difficult to see or find?
• If you need to compare or combine different parts, can you see them at the same time? If not, why not?

### Diffuseness
• Does the notation a) let you say what you want reasonably briefly, or b) is it long-winded? Why?
• What sorts of things take more space to describe?

### Hard Mental Operations
• What kind of things require the most mental effort with this notation?
• Do some things seem especially complex or difficult to work out in your head (e.g. when combining several things)?
• What are they?

### Error Proneness
• Do some kinds of mistake seem particularly common or easy to make? Which ones?
• Do you often find yourself making small slips that irritate you or make you feel stupid? What are some examples?

### Closeness of Mapping
• How closely related is the notation to the result that you are describing? Why? (Note that in a sub-device, the result
may be part of another notation, rather than the end product).
• Which parts seem to be a particularly strange way of doing or describing something?

### Role Expressiveness
• When reading the notation, is it easy to tell what each part is for in the overall scheme? Why?
• Are there some parts that are particularly difficult to interpret? Which ones?
• Are there parts that you really don’t know what they mean, but you put them in just because it’s always been that way? What are they?

### Progressive Evaluation
• How easy is it to stop in the middle of creating some notation, and check your work so far? Can you do this any time you like? If not, why not?
• Can you find out how much progress you have made, or check what stage in your work you are up to? If not, why not?
• Can you try out partially-completed versions of the product? If not, why not?

### Provisionality
• Is it possible to sketch things out when you are playing around with ideas, or when you aren’t sure which way to proceed? What features of the notation help you to do this?
• What sort of things can you do when you don’t want to be too precise about the exact result you are trying to get?

### Premature Commitment
• When you are working with the notation, can you go about the job in any order you like, or does the system force you to think ahead and make certain decisions first?
• If so, what decisions do you need to make in advance? What sort of problems can this cause in your work?

### Consistency
• Where there are different parts of the notation that mean similar things, is the similarity clear from the way they appear? Please give examples.
• Are there places where some things ought to be similar, but the notation makes them different? What are they?

### Secondary Notation
• Is it possible to make notes to yourself, or express information that is not really recognised as part of the notation?
• If it was printed on a piece of paper that you could annotate or scribble on, what would you write or draw?
• Do you ever add extra marks (or colours or format choices) to clarify, emphasise or repeat what is there already? [If yes: does this constitute a helper device? If so, please fill in one of the section 5 sheets describing it]

### Abstraction Management
• Does the system give you any way of defining new facilities or terms within the notation, so that you can extend it to describe new things or to express your ideas more clearly or succinctly? What are they?
• Does the system insist that you start by defining new terms before you can do anything else? What sort of things?

**Comment by eernstg** [comment]
Very nice list!  I've added a couple of comments inline, mainly about words
that it might be useful to include because (according to the experience I
have from a programming language design & foundations context) those words
would be used quite frequently when talking about the given concepts, and
hence they might be a bit more recognizable / self-explanatory.

On Sat, Mar 7, 2015 at 1:35 PM, Luke Church <notifications@github.com>
wrote:

> @Emasoft <https://github.com/Emasoft> @munificent
> <https://github.com/munificent> @sethladd <https://github.com/sethladd>
> @gbracha <https://github.com/gbracha>
>
> I love the idea of adding a usability analysis to proposals. I think the
> most important part of this is that it would mandate that people supply an
> argument justifying their proposal in terms of usability, rather than
> seeking a number.
>
> The system being proposed here is known in UX circles as a form of
> 'heuristic evaluation'. Such heuristics themselves are hard to come up with
> and evaluate for efficacy and completeness. The scheme proposed here
> focuses primarily on readability, important certainly - but not sufficient.
>
> Even though we don't have a mathematical model for the brain, we don't
> need one to, as you say there are other disciplines like psychology that
> can help us here. There are people who have spent a couple of decades
> thinking about what makes notation systems usable, and some of their models
> have gone through some basic cross-validation.
>
> http://www.ppig.org/papers/12th-blackwell.pdf, (snippits extracted at the
> bottom) is probably the most immediately relevant set of questions. This is
> a scheme I've worked with extensively on other programming design projects
> and have generally found it very helpful.
>
> My suggestion is that in order to be considered a proposal must include a
> 'usability analysis', which justifies the UX implications of the change. It
> is up to the authors of the proposal what they think is the most persuasive
> way of demonstrating UX benefits, however I am most likely to be persuaded
> of completeness of a UX proposal if it has answers to the below for the
> proposal. For some proposals, some of those answers may be "this needs tool
> support to do x", or "not applicable here".
>
> If the proposal proceeds to prototype, in many cases it's reasonable to
> run empirical studies to determine whether it has e.g. readability problems
> or not. I'd rather we used validated heuristics for the proposal and
> focused on getting numbers measured from studies.
>
> This may seem like a lot of work, but I think it's a fairly reasonable
> amount given the complexity of assessing the user impact of changes to a
> programming language. The questionnaire below often takes around an hour or
> so to fill out. Given that a change would [hopefully] positively improve
> every Dart developer's UX, that seems a fairly small cost.
>
> Thanks for bringing this up, I think it's a very important aspect of the
> process.
> Viscosity
>
> • When you need to make changes to previous work, how easy is it to make
> the change? Why?
> • Are there particular changes that are more difficult or especially
> difficult to make? Which ones?
>
Software with low viscosity would often be called 'malleable' software, and
references to 'maintainability', 'customizability', or 'extensibility'
would be quite typical.  By the way, 'viscosity' is a quite nice term for
the concept, I just haven't heard it used in this context before, so maybe
many readers with a similar background would be confused and skip to the
next item.

> Hidden Dependencies
>
> • If the structure of the proposal means some parts are closely related to
> other parts, and changes to one may affect the other, are those
> dependencies visible? What kind of dependencies are hidden?
> • In what ways can it get worse when you are creating a particularly large
> description?
> • Do these dependencies stay the same, or are there some actions that
> cause them to get frozen? If so, what are they?
>
For this, I'm more used to a distinction between 'explicit' and 'implicit'
dependencies.  An explicit dependency is a syntactic element that denotes
(typically: uses the name of) the entity depended upon, at some granularity
(maybe we just depend on a library, maybe we depend on specific entities
defined therein, maybe only in a restricted manner like "client-of" as
opposed to "inheriting-from", etc.etc.).

> Visibility and Juxtaposability
>
> • How easy is it to see or find the various parts of the notation while it
> is being created or changed? Why?
> • What kind of things are more difficult to see or find?
> • If you need to compare or combine different parts, can you see them at
> the same time? If not, why not?
>
Given that this is a language issue and not a tooling issue, do we have any
other degrees of visibility than "in this file" and "in an imported library
(possibly via exports)"?  Is this an important dimension?

> Diffuseness
>
> • Does the notation a) let you say what you want reasonably briefly, or b)
> is it long-winded? Why?
> • What sorts of things take more space to describe?
>
"Conciseness"?

> Hard Mental Operations
>
> • What kind of things require the most mental effort with this notation?
> • Do some things seem especially complex or difficult to work out in your
> head (e.g. when combining several things)?
> • What are they?
>
Nice!

> Error Proneness
>
> • Do some kinds of mistake seem particularly common or easy to make? Which
> ones?
> • Do you often find yourself making small slips that irritate you or make
> you feel stupid? What are some examples?
>
Nice, too!

> Closeness of Mapping
>
> • How closely related is the notation to the result that you are
> describing? Why? (Note that in a sub-device, the result
> may be part of another notation, rather than the end product).
> • Which parts seem to be a particularly strange way of doing or describing
> something?
>
This seems to be an issue along the lines of "does the notation resemble
the semantics?".  One well-known topic related to this is
"compositionality":  That it is (to some degree) possible to specify the
semantics of a composite entity by taking the semantics of its syntactic
parts and composing those semantics ("semanticses"? ;-) in a systematic
manner.

> Role Expressiveness
>
> • When reading the notation, is it easy to tell what each part is for in
> the overall scheme? Why?
> • Are there some parts that are particularly difficult to interpret? Which
> ones?
> • Are there parts that you really don’t know what they mean, but you put
> them in just because it’s always been that way? What are they?
>
Interestingly, you would expect high role expressiveness for compositional
mechanisms, so maybe compositionality covers both, to some extent.

> Progressive Evaluation
>
> • How easy is it to stop in the middle of creating some notation, and
> check your work so far? Can you do this any time you like? If not, why not?
> • Can you find out how much progress you have made, or check what stage in
> your work you are up to? If not, why not?
> • Can you try out partially-completed versions of the product? If not, why
> not?
>
Nice concept!

> Provisionality
>
> • Is it possible to sketch things out when you are playing around with
> ideas, or when you aren’t sure which way to proceed? What features of the
> notation help you to do this?
> • What sort of things can you do when you don’t want to be too precise
> about the exact result you are trying to get?
> Premature Commitment
>
> • When you are working with the notation, can you go about the job in any
> order you like, or does the system force you to think ahead and make
> certain decisions first?
> • If so, what decisions do you need to make in advance? What sort of
> problems can this cause in your work?
>
'The system' sounds like tooling, which would presumably not be the topic
in this context.

> Consistency
>
> • Where there are different parts of the notation that mean similar
> things, is the similarity clear from the way they appear? Please give
> examples.
> • Are there places where some things ought to be similar, but the notation
> makes them different? What are they?
>
Nice checks!

> Secondary Notation
>
> • Is it possible to make notes to yourself, or express information that is
> not really recognised as part of the notation?
> • If it was printed on a piece of paper that you could annotate or
> scribble on, what would you write or draw?
> • Do you ever add extra marks (or colours or format choices) to clarify,
> emphasise or repeat what is there already? [If yes: does this constitute a
> helper device? If so, please fill in one of the section 5 sheets describing
> it]
>
This sounds like comments.  Do we need to ask such a question for every
proposed language enhancement?

> Abstraction Management
>
> • Does the system give you any way of defining new facilities or terms
> within the notation, so that you can extend it to describe new things or to
> express your ideas more clearly or succinctly? What are they?
>
This might be somewhat trivial:  For a language construct it would be
obvious whether it defines a new entity (with a name, or associated with
some other mechanism that allows for "application" of the entity zero or
more times).

> • Does the system insist that you start by defining new terms before you
> can do anything else? What sort of things?
>
Toolish, again.  But of course, the issue could be whether a program can be
compiled and executed with a partial specification of some (planned,
intended) complete software modification/extension using the proposed
language enhancement.

  best regards,

-- 
Erik Ernst  -  Google Danmark ApS
Skt Petri Passage 5, 2 sal, 1165 København K, Denmark
CVR no. 28866984

**Comment by munificent** [comment]
I think usability is hugely important, and I think we should think about usability considerations throughout the entire language evolution process.

At the same time, I don't want us to do that in overly formal way. Usability isn't a simple game where you can calculate a score and whoever has the most points is the most usable. It requires care and thought and human judgement.

So, I want to keep the sentiment here, but I'm closing this because, if anything, I'd like our process to get less formal, not more.
